OverlayPlugin.Core/Integration/FFXIVCustomLogLines.cs

```diff
@@ -0,0 +1,182 @@
+                        {
+                            if (registry.ContainsKey(ID))
+                            {
+                                logger.Log(LogLevel.Warning, $"Reserved log line entry already registered ({ID}).");
+                                continue;
+                            }
+                            var Source = reservedDataEntry.Source;
+                            var Version = reservedDataEntry.Version.Value;
+                            logger.Log(LogLevel.Debug, $"Reserving log line entry for ID {ID}, Source {Source}, Version {Version}.");
+                            registry[ID] = new LogLineRegistryEntry()
+                            {
+                                ID = ID,
+                                Source = Source,
+                                Version = Version,
+                            };
+                        }
+                    }
+                    else
+                    {
+                        var ID = reservedDataEntry.ID.Value;
+                        if (registry.ContainsKey(ID))
+                        {
+                            logger.Log(LogLevel.Warning, $"Reserved log line entry already registered ({ID}).");
+                            continue;
+                        }
+                        var Source = reservedDataEntry.Source;
+                        var Version = reservedDataEntry.Version.Value;
+                        logger.Log(LogLevel.Debug, $"Reserving log line entry for ID {ID}, Source {Source}, Version {Version}.");
+                        registry[ID] = new LogLineRegistryEntry()
+                        {
+                            ID = ID,
+                            Source = Source,
+                            Version = Version,
+                        };
+                    }
+                }
+            } catch(Exception ex)
+            {
+                logger.Log(LogLevel.Error, string.Format(Resources.ErrorCouldNotLoadReservedLogLines, ex));
+            }
+        }
+
+        public Func<string, bool> RegisterCustomLogLine(ILogLineRegistryEntry entry)
+        {
+            // Don't allow any attempt to write a custom log line with FFXIV_ACT_Plugin as the source.
+            // This prevents a downstream plugin from attempting to register e.g. `00` lines by just pretending to be FFXIV_ACT_Plugin.
+            if (entry.Source == "FFXIV_ACT_Plugin")
+            {
+                logger.Log(LogLevel.Warning, $"Attempted to register custom log line with reserved source.");
+                return null;
+            }
+            var ID = entry.ID;
+            if (registry.ContainsKey(ID))
+            {
+                // Allow re-registering the handler if the ID, Source, and Version match.
+                // Implicitly don't allow re-registering the same handler if the Version changes to prevent log file confusion.

```

Comment:
I'm not sure I understand what this is trying to prevent.  I could understand trying to prevent two plugins from registering the same id (regardless of version).  But this is just looking for agreement between the json version and the plugin providing it.  I think if somehow a user updates OverlayPlugin and gets a new json version but doesn't update some non-OverlayPlugin plugin that's providing the log line, that shouldn't be an error.  The registration line should print out what version it is.

Reply:
Agreed, I didn't think about the out-of-sync version updating aspect. Removed `Version` from the check.

Reply:
So, a couple of minor thoughts that shouldn't block merging this either.  Stop me if I'm overthinking edge cases.

(1) Without a version check at all, it means you can register custom lines with any id, even ones that aren't reserved specifically.  (Maybe that's ok, but it seems nice to have a "you must PR what this line type is before using it" restriction, e.g. you can only register explicit lines that are reserved at Version>=1 and not the large pre-reserved block Version==0? Or something along those lines)

(2) If two plugins somehow registered the same log line id but different versions, they'd emit different info for it.  (Maybe once something is explicitly registered, we don't allow further explicit registrations?  I think even two plugins with the same id/source/version should probably both not emit the same thing.

Reply:
1. I didn't want to require a PR to use the functionality because that would block people from doing dev work faster, though I guess they could always edit their local json as needed. I think it does make sense to require a `Version` of `>0` regardless of registration status, essentially making `0` a config-file-only thing?
2. The two plugins would have to register the same `ID` and the same `Source` for it to be allowed. There's still an overridden `Equals` check here which checks for `Source`+`ID`.

Maybe the logic isn't 100% clear I guess? The logic here works out to:

1. If a line ID was pre-reserved in the JSON, allow registering if the `ID` and `Source` match. This allows the pre-reservation of IDs.
2. If a line ID has already been registered previously, allow registering if the `ID` and `Source` match. This allows plugins to be e.g. unloaded and reloaded and still be able to register handlers.

Reply:
All of these explanations make sense to me.  Version=0 being config file sounds good.  I hadn't thought about the unloading and reloading.  (I feel like that works so little for me that I just restart ACT if I ever have to unload anything, and don't even consider that use case.)